File: crates/mithras-crypto/src/address.rs
use anyhow::{Result, anyhow};
use bech32::Hrp;
use ed25519_dalek::VerifyingKey as Ed25519PublicKey;
use serde::{Deserialize, Serialize};
use x25519_dalek::PublicKey as X25519PublicKey;

use crate::{
    MithrasError,
    hpke::{SupportedHpkeSuite, SupportedNetwork},
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MithrasAddr {
    pub version: u8,
    pub network: SupportedNetwork,
    pub suite: SupportedHpkeSuite,
    pub spend_ed25519: Ed25519PublicKey,
    pub disc_x25519: X25519PublicKey,
}

impl MithrasAddr {
    pub fn encode(&self) -> Result<String, MithrasError> {
        let mut data = Vec::<u8>::with_capacity(3 + 32 + 32);
        data.push(self.version);
        data.push(self.network.into());
        data.push(self.suite.into());
        data.extend_from_slice(&self.spend_ed25519.to_bytes());
        data.extend_from_slice(&self.disc_x25519.to_bytes());

        bech32::encode::<bech32::Bech32m>(Hrp::parse_unchecked("mith"), &data)
            .map_err(|e| MithrasError::AddressEncoding { msg: e.to_string() })
    }

    pub fn decode(s: &str) -> Result<Self> {
        let (hrp, data) = bech32::decode(s)?;
        if hrp.as_str() != "mith" {
            return Err(anyhow!(
                "invalid human-readable prefix. Got {}, expected mith",
                hrp
            ));
        }
        let version = data[0];
        let network = data[1];
        let suite = data[2];
        let spend = data[3..35].try_into()?;
        let disc_arr: [u8; 32] = data[35..67].try_into()?;
        let disc = X25519PublicKey::from(disc_arr);

        Ok(Self {
            version,
            network: network.try_into()?,
            suite: suite.try_into()?,
            spend_ed25519: spend,
            disc_x25519: disc,
        })
    }

    pub fn from_keys(
        spend: &Ed25519PublicKey,
        disc: &X25519PublicKey,
        version: u8,
        network: SupportedNetwork,
        suite: SupportedHpkeSuite,
    ) -> Self {
        Self {
            version,
            network,
            suite,
            spend_ed25519: *spend,
            disc_x25519: *disc,
        }
    }
}

File: packages/mithras-crypto/src/address.ts
import { bech32 } from "bech32";

export class MithrasAddr {
  version: number;
  network: number;
  suite: number;
  spendEd25519: Uint8Array;
  viewX25519: Uint8Array;

  constructor(
    version: number,
    network: number,
    suite: number,
    spendEd25519: Uint8Array,
    viewX25519: Uint8Array,
  ) {
    this.version = version;
    this.network = network;
    this.suite = suite;

    if (spendEd25519.length !== 32) {
      throw new Error(
        `invalid spendEd25519 length. Got ${spendEd25519.length}, expected 32`,
      );
    }
    if (viewX25519.length !== 32) {
      throw new Error(
        `invalid viewX25519 length. Got ${viewX25519.length}, expected 32`,
      );
    }

    this.spendEd25519 = spendEd25519;
    this.viewX25519 = viewX25519;
  }

  static fromKeys(
    spend: Uint8Array,
    view: Uint8Array,
    version: number,
    network: number,
    suite: number,
  ): MithrasAddr {
    return new MithrasAddr(version, network, suite, spend, view);
  }

  encode(): string {
    const data = new Uint8Array(3 + 32 + 32);
    data[0] = this.version;
    data[1] = this.network;
    data[2] = this.suite;
    data.set(this.spendEd25519, 3);
    data.set(this.viewX25519, 35);

    return bech32.encode("mith", bech32.toWords(data), 200);
  }

  static decode(s: string): MithrasAddr {
    const { prefix, words } = bech32.decode(s, 200);
    if (prefix !== "mith") {
      throw new Error(
        `invalid human-readable prefix. Got ${prefix}, expected mith`,
      );
    }
    const data = new Uint8Array(bech32.fromWords(words));
    const version = data[0];
    const network = data[1];
    const suite = data[2];
    const spend = data.slice(3, 35);
    const view = data.slice(35, 67);

    return new MithrasAddr(version, network, suite, spend, view);
  }
}
